### ISSUE TITLE
Support "whitelistPackages" for faster scanning

### ISSUE THREAD

**Issue by antch** [vojtechhabarta/typescript-generator#330]
As an experiment I used `ClassGraph#whitelistPackages()` to limit the packages that should be scanned for my project. This cut the runtime of typescript-generator from 22 seconds down to 15. It would be extremely helpful for this configuration option to be supported.

**Comment by antch** [comment]
I have this enhancement ready to go if you're on board with it.

**Comment by vojtechhabarta** [comment]
I am reluctant to add this feature for following reasons:
- users could configure whitelist inconsistently with `classPatterns` parameter which could be hard to debug
- each configuration parameter creates clutter and makes the usage less straightforward

On the other hand I agree that 15 vs. 22 seconds is significant improvement.

**Comment by antch** [comment]
IMO any conflict with `classPatterns` would be pretty obvious. Where I see potential confusion is that explicitly-specified "classes" would not be subject to the whitelist (since that does not use the `Classgraph` object). My PR does specifically document that limitation FWIW.

Is there any other way aside from adding an explicit parameter, perhaps some mechanism to allow arbitrary customization or factory for the `Classgraph` instance? Though, now that I type it out that seems like a nightmare.

I am looking to break my usage down into three different executions of the generator, one for each superclass, and without the whitelisting the runtime would make it a non-starter...

**Comment by antch** [comment]
I have a cleaned-up branch ready to go, definitely willing to work with you on any changes if you would consider the feature!

**Comment by vojtechhabarta** [comment]
I checked some build server logs of 10 different services. Numbers of scanned classes were from 22k to 36k and scanning times were from 0.5 to 2.2 seconds. On my local machine 36k classes were scanned in 4.2 seconds which is slower but still it's not something which would require optimization.

So I am curious why it is so slow in your case. How many classes are scanned in your run? Or maybe there is some difference in caching, in my case typescript-generator is run as part of maven build so needed files may already be in file system cache.

What do you think? This would be really ideal if we could have fast runs without requiring any additional parameter.

**Comment by antch** [comment]
My project is undergoing a Maven restructuring, so I wanted to wait until I could test this again as I realized I wasn't even scanning all of the classes that I needed. The difference is still ~7 seconds.

With whitelisting:
```
Scanning finished in 11.12 seconds. Total number of classes: 189788.
```

Without whitelisting:
```
Scanning finished in 3.91 seconds. Total number of classes: 28778.
```

**Comment by vojtechhabarta** [comment]
Your times seems to be consistent with my case. 190k classes is big number so the speed is about the same.

I would consider adding parameter for classpath scanning whitelisting but I would like it to have the same syntax as existing parameters (`classPatterns`, `excludeClassPatterns`, `mapClassesAsClassesPatterns`) which is probably not possible to achieve with current classgraph version. (We would need some whitelist filter function.)

So I at least optimized scanning. Instead of `new ClassGraph().enableAllInfo()` only needed info is now enabled. This shortens classpath scanning time by about 20% for all users for free (without configuring anything).

**Comment by antch** [comment]
I am a bit confused on what you mean by "not possible to achieve with the current Classgraph version"? The `whitelistPackages` already exists on the Classgraph object and the parameter can be named anything for the Maven plugin.

If you let me know what naming to use (`classesWithinPackages`, etc) I can have a PR ready within the day.

**Comment by vojtechhabarta** [comment]
Thanks for testing my change, it is good to know the impact.

Regarding naming consistency: it is not needed to have name consistent with ClassGraph because this library might be replaced by other classpath scanning library in the future.

What I meant was parameter named for example `classpathScanningWhitelistPatterns` where I could specify for example `cz.habarta.example.**` or even `**.data.*`. This would be consistent with other typescript-generator parameters.

I would not do anything with this issue now but I would leave it opened.

**Comment by antch** [comment]
I'm not sure how I missed it in the GlassGraph documentation but it does appear that `whitelistPackages` supports `*` as a glob pattern. It does not support `**`, the single asterisk can be used the same way. For example `com.*.dto` would cause the `com.mycompany.mypackage.dto` package to be scanned.

**Comment by vojtechhabarta** [comment]
Commit "Classpath scanning optimization, classgraph updated to 4.8.24" is now released in [2.14.505](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.14.505).

**Comment by antch** [comment]
Please let me know if you are open to any variant of this feature, or anything that allows me to hook into the library to achieve it. I am willing to help out with implementation.

**Comment by vojtechhabarta** [comment]
I committed (3c25020825a55b86091e2dd75c7c794a98b4136c) `Input` class refactoring so it is easier to add parameter in a compatible way.

How do you use typescript-generator? Using Maven or Gradle plugins or direct Java call?

**Comment by antch** [comment]
I am using the Maven plugin, possibly Gradle in the future. Thanks!

**Comment by antch** [comment]
I know this is a bit of an old issue but I was wondering if you had given this any more thought. The generated TS file in my main project is at almost 1MB, and I can't break it up into multiple files unless `whitelistPackages` is supported as the runtime is just too long.

**Comment by vojtechhabarta** [comment]
Yes, I am considering adding support for this parameter.

**Comment by vojtechhabarta** [comment]
I added `scanningAcceptedPackages` parameter (roughly following naming in ClassGraph).

**Comment by antch** [comment]
Awesome, thank you! I will test this out in my project and let you know how it works out. I'm wondering if you can publish the latest version of the Maven plugin since it's several versions behind on Maven Central.

**Comment by vojtechhabarta** [comment]
Tags are created automatically by each build run but proper release (with description, uploaded to Maven Central repo) is done manually (about once a month).

**Comment by vojtechhabarta** [comment]
Released in [2.29.814](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.29.814).

**Comment by antch** [comment]
Thank you, works great!
